pkg/controllers/allocation/batch.go

```diff
@@ -0,0 +1,150 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package allocation
+
+import (
+	"sync"
+	"time"
+
+	"github.com/patrickmn/go-cache"
+)
+
+const (
+	BatchCacheTTL             = 1 * time.Hour
```

Comment:
Why do we need to evict from the cache? Is it just because if the Provisioner goes away then you no longer need the batcher? (Seems like there might be a cleaner way to do that?)

Reply:
Yes, I didn't want to keep them around forever if there was a long running installation with a lot of provisioners (created and delete), it could be storing a lot of batches. I don't think this would be a huge problem in reality, so I opted for the cache as a free-ish way to do the evictions based on time not used.   

An alternative would be to process Provisioner deletions in the reconcile and remove the batch for the provisioner. That would be really simple and explicit.

Reply:
After thinking about it some more, I would suggest something along these lines ... 

1. Have a single background goroutine that wakes up every N seconds (it will be used by all Provisioners)
2. Provisioners start new batches by sending messages to it on a channel
3. Every time the Batcher wakes up it goes thru all the currently active batches and does <<whatever>> with them (fire them off if they're ready, extend if not, deal with additions, yadda yadda)
4. Just punt on cleanup - a single goroutine in the bg isn't a big deal

Reply:
👍  I refactored to get rid of the locks , let me know what you think of it.